livekit-agents/livekit/agents/transcription/tts_forwarder.py

```diff
@@ -0,0 +1,232 @@
+from collections import deque
+import math
+import time
+from attrs import define
+import asyncio
+import contextlib
+import uuid
+from typing import Callable, AsyncIterable
+
+from attrs import define
+from livekit import rtc
+
+from .. import stt, tokenize, tts
+from ..log import logger
+
+
+@define
+class TTSOptions:
+    room: rtc.Room
+    participant_identity: str
+    track_id: str
+    language: str
+    speed: float
+    word_tokenizer: tokenize.WordTokenizer
+    sentence_tokenizer: tokenize.SentenceTokenizer
+    hyphenate_word: Callable[[str], list[str]]
+
+
+@define
+class _SegmentData:
+    sentence_stream: tokenize.SentenceStream
+    text: str
+    audio_duration: float
+    avg_speed: float | None
+    processed_hyphenes: int
+
+
+@define
+class _PendingSegments:
+    cur_audio: _SegmentData
+    cur_text: _SegmentData
+    q: deque[_SegmentData]
+
+
+def _uuid() -> str:
+    return str(uuid.uuid4())[:12]
+
+
+class TTSSegmentsForwarder:
+    def __init__(
+        self,
+        opts: TTSOptions,
+    ):
+        self._opts = opts
+        self._main_task = asyncio.create_task(self._run())
+
+        # current segment where the user may still be pushing text & audio
+        first_segment = self._create_segment()
+        segments_q = deque()
+        segments_q.append(first_segment)
+
+        self._pending_segment = _PendingSegments(
+            cur_audio=first_segment,
+            cur_text=first_segment,
+            q=segments_q,
+        )
+
+        self._seg_queue = asyncio.Queue[_SegmentData | None]()
+        self._seg_queue.put_nowait(first_segment)
+
+        # the forwarding of the transcription is started as soon as the user push the first audio frame.
+        self._start_future = asyncio.Future()
+
+    def validate(self) -> None:
+        """Validate must be called once we started to playout the audio to the user"""
```

Comment:
mb start is a better name?

Reply:
Wondering if we need this at all? Under what conditions would we push the first audio frame and not want to call validate?

Reply:
same question as neil.. I'm not sure why we would make a user call a required function?